File: app/src/main/java/com/genonbeta/TrebleShot/activity/WelcomeActivity.java
package com.genonbeta.TrebleShot.activity;

import android.content.Intent;
import android.content.res.ColorStateList;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.view.animation.OvershootInterpolator;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.drawable.DrawableCompat;
import androidx.transition.TransitionManager;
import androidx.viewpager.widget.ViewPager;

import com.genonbeta.TrebleShot.R;
import com.genonbeta.TrebleShot.app.Activity;
import com.genonbeta.TrebleShot.object.NetworkDevice;
import com.genonbeta.TrebleShot.util.AppUtils;
import com.genonbeta.TrebleShot.widget.DynamicViewPagerAdapter;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

public class WelcomeActivity extends Activity
{
    public final static String TAG = WelcomeActivity.class.getSimpleName();

    private ViewGroup mSplashView;
    private ViewGroup mProfileView;
    private ViewGroup mPermissionsView;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_welcome);
        setSkipPermissionRequest(true);
        setWelcomePageDisallowed(true);

        final FloatingActionButton nextButton = findViewById(R.id.activity_welcome_view_next);
        final AppCompatImageView previousButton = findViewById(R.id.activity_welcome_view_previous);
        final ProgressBar progressBar = findViewById(R.id.activity_welcome_progress_bar);
        final ViewPager viewPager = findViewById(R.id.activity_welcome_view_pager);
        final DynamicViewPagerAdapter pagerAdapter = new DynamicViewPagerAdapter();

        {
            @ColorInt
            int appliedColor = ContextCompat.getColor(this, AppUtils.getReference(this, R.attr.colorSecondary));

            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
                Drawable wrapDrawable = DrawableCompat.wrap(progressBar.getProgressDrawable());

                DrawableCompat.setTint(wrapDrawable, appliedColor);
                progressBar.setProgressDrawable(DrawableCompat.unwrap(wrapDrawable));
            } else
                progressBar.setProgressTintList(ColorStateList.valueOf(appliedColor));
        }

        {
            mSplashView = (ViewGroup) getLayoutInflater().inflate(R.layout.layout_welcome_page_1, null, false);
            pagerAdapter.addView(mSplashView);
        }

        if (Build.VERSION.SDK_INT >= 23) {
            mPermissionsView = (ViewGroup) getLayoutInflater().inflate(R.layout.layout_welcome_page_3, null, false);
            pagerAdapter.addView(mPermissionsView);
            checkPermissionsState();

            mPermissionsView.findViewById(R.id.layout_welcome_page_3_request_button)
                    .setOnClickListener(new View.OnClickListener()
                    {
                        @Override
                        public void onClick(View v)
                        {
                            requestRequiredPermissions(false);
                        }
                    });
        }

        {
            mProfileView = (ViewGroup) getLayoutInflater().inflate(R.layout.layout_welcome_page_2, null, false);
            pagerAdapter.addView(mProfileView);
            setUserProfile();
        }

        pagerAdapter.addView(getLayoutInflater().inflate(R.layout.layout_welcome_page_4, null, false));

        {
            View view = getLayoutInflater().inflate(R.layout.layout_welcome_page_5, null, false);
            AlphaAnimation alphaAnimation = new AlphaAnimation(0.3f, 1.0f);

            alphaAnimation.setDuration(2000);
            alphaAnimation.setRepeatCount(Animation.INFINITE);
            alphaAnimation.setRepeatMode(Animation.REVERSE);

            view.findViewById(R.id.layout_welcome_page_5_text)
                    .setAnimation(alphaAnimation);

            pagerAdapter.addView(view);
        }
        progressBar.setMax((pagerAdapter.getCount() - 1) * 100);

        previousButton.setOnClickListener(new View.OnClickListener()
        {
            @Override
            public void onClick(View v)
            {
                if (viewPager.getCurrentItem() - 1 >= 0)
                    viewPager.setCurrentItem(viewPager.getCurrentItem() - 1, true);
            }
        });

        nextButton.setOnClickListener(new View.OnClickListener()
        {
            @Override
            public void onClick(View v)
            {
                if (viewPager.getCurrentItem() + 1 < pagerAdapter.getCount())
                    viewPager.setCurrentItem(viewPager.getCurrentItem() + 1);
                else {
                    // end presentation
                    getDefaultPreferences().edit()
                            .putBoolean("introduction_shown", true)
                            .apply();

                    startActivity(new Intent(WelcomeActivity.this, HomeActivity.class));
                    finish();
                }
            }
        });

        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener()
        {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels)
            {
                progressBar.setProgress((position * 100) + (int) (positionOffset * 100));

                if (position == 0) {
                    progressBar.setAlpha(positionOffset);
                    previousButton.setAlpha(positionOffset);
                } else {
                    progressBar.setAlpha(1.0f);
                    previousButton.setAlpha(1.0f);
                }
            }

            @Override
            public void onPageSelected(int position)
            {
                OvershootInterpolator interpolator = new OvershootInterpolator();

                nextButton.setImageResource(position + 1 >= pagerAdapter.getCount()
                        ? R.drawable.ic_check_white_24dp
                        : R.drawable.ic_navigate_next_white_24dp);
            }

            @Override
            public void onPageScrollStateChanged(int state)
            {
            }
        });

        viewPager.setAdapter(pagerAdapter);
    }

    @Override
    protected void onResume()
    {
        super.onResume();

        slideSplashView();
        setUserProfile();
        checkPermissionsState();
    }

    @Override
    public void onUserProfileUpdated()
    {
        super.onUserProfileUpdated();
        setUserProfile();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults)
    {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        checkPermissionsState();
    }

    protected void checkPermissionsState()
    {
        if (Build.VERSION.SDK_INT < 23)
            return;

        boolean permissionsOk = AppUtils.checkRunningConditions(this);

        mPermissionsView.findViewById(R.id.layout_welcome_page_3_perm_ok_image)
                .setVisibility(permissionsOk ? View.VISIBLE : View.GONE);

        mPermissionsView.findViewById(R.id.layout_welcome_page_3_request_button)
                .setVisibility(permissionsOk ? View.GONE : View.VISIBLE);
    }

    protected void setUserProfile()
    {
        if (mProfileView != null) {
            NetworkDevice localDevice = AppUtils.getLocalDevice(getApplicationContext());

            ImageView imageView = mProfileView.findViewById(R.id.layout_profile_picture_image_default);
            ImageView editImageView = mProfileView.findViewById(R.id.layout_profile_picture_image_preferred);
            TextView deviceNameText = mProfileView.findViewById(R.id.header_default_device_name_text);
            TextView versionText = mProfileView.findViewById(R.id.header_default_device_version_text);

            deviceNameText.setText(localDevice.nickname);
            versionText.setText(localDevice.versionName);
            loadProfilePictureInto(localDevice.nickname, imageView);

            editImageView.setOnClickListener(new View.OnClickListener()
            {
                @Override
                public void onClick(View v)
                {
                    startProfileEditor();
                }
            });

            TransitionManager.beginDelayedTransition(mProfileView);
        }
    }

    protected void slideSplashView()
    {
        mSplashView.findViewById(R.id.layout_welcome_page_1_splash_image)
                .setAnimation(AnimationUtils.loadAnimation(this, R.anim.enter_from_bottom_centered));

        mSplashView.findViewById(R.id.layout_welcome_page_1_details)
                .setAnimation(AnimationUtils.loadAnimation(this, R.anim.enter_from_bottom));
    }
}

File: app/src/main/java/com/genonbeta/TrebleShot/activity/WelcomeActivity.kt
/*
 * Copyright (C) 2019 Veli Tasalı
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package com.genonbeta.TrebleShot.activity

import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.drawable.Drawable
import android.os.*
import android.view.View
import android.view.ViewGroup
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.appcompat.widget.AppCompatImageView
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.DrawableCompat
import androidx.transition.TransitionManager
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.genonbeta.TrebleShot.R
import com.genonbeta.TrebleShot.app.Activity
import com.genonbeta.TrebleShot.util.AppUtils
import com.genonbeta.TrebleShot.widget.DynamicViewPagerAdapter
import com.google.android.material.floatingactionbutton.FloatingActionButton

class WelcomeActivity : Activity() {
    private lateinit var splashView: ViewGroup

    private lateinit var profileView: ViewGroup

    private lateinit var permissionsView: ViewGroup

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (hasIntroductionShown()) finish()
        setContentView(R.layout.activity_welcome)

        skipPermissionRequest = true
        welcomePageDisallowed = true

        val nextButton: FloatingActionButton = findViewById(R.id.activity_welcome_view_next)
        val previousButton: AppCompatImageView = findViewById(R.id.activity_welcome_view_previous)
        val progressBar = findViewById<ProgressBar>(R.id.activity_welcome_progress_bar)
        val viewPager: ViewPager = findViewById(R.id.activity_welcome_view_pager)
        val pagerAdapter = DynamicViewPagerAdapter()

        @ColorInt val appliedColor: Int = ContextCompat.getColor(
            this, AppUtils.getReference(
                this,
                R.attr.colorSecondary
            )
        )
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            val wrapDrawable: Drawable = DrawableCompat.wrap(progressBar.progressDrawable)
            DrawableCompat.setTint(wrapDrawable, appliedColor)
            progressBar.progressDrawable = DrawableCompat.unwrap(wrapDrawable)
        } else progressBar.progressTintList = ColorStateList.valueOf(appliedColor)

        layoutInflater.inflate(R.layout.layout_welcome_page_1, null, false).also {
            pagerAdapter.addView(it)
            splashView = it as ViewGroup
        }

        if (Build.VERSION.SDK_INT >= 23) layoutInflater.inflate(R.layout.layout_welcome_page_3, null, false).also {
            pagerAdapter.addView(it)
            checkPermissionsState()
            it.findViewById<View>(R.id.layout_welcome_page_3_request_button).setOnClickListener {
                requestRequiredPermissions(false)
            }

            permissionsView = it as ViewGroup
        }

        layoutInflater.inflate(R.layout.layout_welcome_page_2, null, false).also {
            pagerAdapter.addView(it)
            setUserProfile()

            profileView = it as ViewGroup
        }

        pagerAdapter.addView(layoutInflater.inflate(R.layout.layout_welcome_page_4, null, false))

        layoutInflater.inflate(R.layout.layout_welcome_page_5, null, false).also { view: View ->
            val alphaAnimation = AlphaAnimation(0.3f, 1.0f)
            alphaAnimation.duration = 2000
            alphaAnimation.repeatCount = Animation.INFINITE
            alphaAnimation.repeatMode = Animation.REVERSE
            view.findViewById<View>(R.id.layout_welcome_page_5_text).animation = alphaAnimation
            pagerAdapter.addView(view)
        }

        progressBar.max = (pagerAdapter.getCount() - 1) * 100

        previousButton.setOnClickListener { v: View? ->
            if (viewPager.currentItem - 1 >= 0) viewPager.setCurrentItem(
                viewPager.currentItem - 1, true
            )
        }

        nextButton.setOnClickListener {
            if (viewPager.currentItem + 1 < pagerAdapter.getCount()) viewPager.currentItem =
                viewPager.getCurrentItem() + 1 else {
                // end presentation
                defaultPreferences.edit()
                    .putBoolean("introduction_shown", true)
                    .apply()
                startActivity(Intent(this@WelcomeActivity, HomeActivity::class.java))
                finish()
            }
        }

        viewPager.addOnPageChangeListener(object : OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
                progressBar.progress = position * 100 + (positionOffset * 100).toInt()
                if (position == 0) {
                    progressBar.alpha = positionOffset
                    previousButton.setAlpha(positionOffset)
                } else {
                    progressBar.alpha = 1.0f
                    previousButton.setAlpha(1.0f)
                }
            }

            override fun onPageSelected(position: Int) {
                nextButton.setImageResource(if (position + 1 >= pagerAdapter.getCount()) R.drawable.ic_check_white_24dp else R.drawable.ic_navigate_next_white_24dp)
            }

            override fun onPageScrollStateChanged(state: Int) {}
        })

        viewPager.setAdapter(pagerAdapter)
    }

    override fun onStart() {
        super.onStart()
        slideSplashView()
    }

    override fun onResume() {
        super.onResume()
        setUserProfile()
        checkPermissionsState()
    }

    override fun onUserProfileUpdated() {
        super.onUserProfileUpdated()
        setUserProfile()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        checkPermissionsState()
    }

    protected fun checkPermissionsState() {
        if (Build.VERSION.SDK_INT < 23) return
        val permissionsOk = AppUtils.checkRunningConditions(this)
        permissionsView.findViewById<View>(R.id.layout_welcome_page_3_perm_ok_image).visibility =
            if (permissionsOk) View.VISIBLE else View.GONE
        permissionsView.findViewById<View>(R.id.layout_welcome_page_3_request_button).visibility =
            if (permissionsOk) View.GONE else View.VISIBLE
    }

    protected fun setUserProfile() {
        val localDevice = AppUtils.getLocalDevice(applicationContext)
        val imageView = profileView.findViewById<ImageView>(R.id.layout_profile_picture_image_default)
        val editImageView = profileView.findViewById<ImageView>(R.id.layout_profile_picture_image_preferred)
        val deviceNameText: TextView = profileView.findViewById(R.id.header_default_device_name_text)
        val versionText: TextView = profileView.findViewById(R.id.header_default_device_version_text)
        deviceNameText.setText(localDevice.username)
        versionText.setText(localDevice.versionName)
        loadProfilePictureInto(localDevice.username, imageView)
        editImageView.setOnClickListener { v: View? -> startProfileEditor() }
        TransitionManager.beginDelayedTransition(profileView)
    }

    protected fun slideSplashView() {
        splashView.findViewById<View>(R.id.layout_welcome_page_1_splash_image).animation =
            AnimationUtils.loadAnimation(this, R.anim.enter_from_bottom_centered)
        splashView.findViewById<View>(R.id.layout_welcome_page_1_details).animation =
            AnimationUtils.loadAnimation(this, R.anim.enter_from_bottom)
    }

    companion object {
        val TAG = WelcomeActivity::class.java.simpleName
    }
}